include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -17,95 +17,303 @@ namespace hlsl
     {
-        MT measure_transform;
-        measure_transform.pdf = ndf.template D<anisocache_type>(cache);
-        NBL_IF_CONSTEXPR(MT::Type == ndf::MicrofacetTransformTypes::MTT_REFLECT_REFRACT)
-            measure_transform.transmitted = cache.isTransmission();
-        NBL_IF_CONSTEXPR(MT::Type == ndf::MicrofacetTransformTypes::MTT_REFRACT)
+        using quant_query_type = typename N::quant_query_type;
+        using g2g1_query_type = typename N::g2g1_query_type;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel.orientedEta;
+        quant_query_type qq = ndf.template createQuantQuery<anisocache_type>(cache, orientedEta.value[0]);
+        g2g1_query_type gq = ndf.template createG2G1Query<sample_type, anisotropic_interaction_type>(_sample, interaction);
+
+        quant_type D = ndf.template D<sample_type, anisotropic_interaction_type, anisocache_type>(qq, _sample, interaction, cache);
+        scalar_type DG = D.projectedLightMeasure;
+        if (any<vector<bool, 2> >(ndf.__base.A > hlsl::promote<vector2_type>(numeric_limits<scalar_type>::min)))
         {
-            measure_transform.VdotH = cache.getVdotH();
-            measure_transform.LdotH = cache.getLdotH();
-            measure_transform.VdotHLdotH = cache.getVdotHLdotH();
+            quant_type G2 = ndf.template correlated<sample_type, anisotropic_interaction_type>(gq, qq, _sample, interaction);
+            DG *= G2.microfacetMeasure;
         }
-        if (any<vector<bool, 2> >(ndf.A > hlsl::promote<vector2_type>(numeric_limits<scalar_type>::min)))
+        return hlsl::promote<spectral_type>(fresnel(hlsl::abs(cache.getVdotH()))[0]) * DG;
+    }
+
+    sample_type generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(isocache_type) cache, NBL_CONST_REF_ARG(spectral_type) luminosityContributionHint)
+    {
+        anisocache_type anisocache;
+        sample_type s = generate(anisotropic_interaction_type::create(interaction), u, anisocache, luminosityContributionHint);
+        cache = anisocache.iso_cache;
+        return s;
+    }
+    sample_type generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(anisocache_type) cache, NBL_CONST_REF_ARG(spectral_type) luminosityContributionHint)
+    {
+        const vector3_type localV = interaction.getTangentSpaceV();
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel.orientedEta;
+        fresnel::OrientedEtaRcps<monochrome_type> rcpEta = orientedEta.getReciprocals();
+
+        const vector3_type upperHemisphereV = hlsl::mix(localV, -localV, interaction.getNdotV() < scalar_type(0.0));
+        const vector3_type H = ndf.generateH(upperHemisphereV, u.xy);
+
+        const scalar_type reflectance = fresnel(hlsl::abs(hlsl::dot(localV, H)))[0];
+        const scalar_type reflectionProb = hlsl::dot<spectral_type>(hlsl::promote<spectral_type>(reflectance), luminosityContributionHint);
```

Comment:
why are you getting and using the R component of the fresnel ?

And later promoting just the R component to later dot with the luminosity contrib hint !? This is surely wrong!

Also please make a `const absVdotH` variable for clarity

Reply:
also `luminosityContributionHint` should be a member just how it is for the thindielectric BxDF

Reply:
fresnel dimensionality is a mess, lots of static asserts needed